src/ec2/model.ts

```diff
@@ -145,11 +167,117 @@ export class Ec2ConnectionManager {
+        const remoteEnv = await this.prepareEc2RemoteEnv(selection, remoteUser).finally(() => timeout.cancel())
+        return remoteEnv
+    }
+
+    public async prepareEc2RemoteEnv(selection: Ec2Selection, remoteUser: string): Promise<Ec2RemoteEnv> {
+        const logger = this.configureRemoteConnectionLogger(selection.instanceId)
+        const { ssm, vsc, ssh } = (await ensureDependencies()).unwrap()
+        const keyPath = await this.configureSshKeys(selection, remoteUser)
+        const sshConfig = new VscodeRemoteSshConfig(ssh, hostNamePrefix, ec2ConnectScriptPrefix, keyPath)
+
+        const config = await sshConfig.ensureValid()
+        if (config.isErr()) {
+            const err = config.err()
+            getLogger().error(`ec2: failed to add ssh config section: ${err.message}`)
+
+            throw err
+        }
+        const session = await this.ssmClient.startSession(selection.instanceId, 'AWS-StartSSHSession')
+        const vars = getEc2SsmEnv(selection, ssm, session)
+        const envProvider = async () => {
+            return { [sshAgentSocketVariable]: await startSshAgent(), ...vars }
+        }
+        const SessionProcess = createBoundProcess(envProvider).extend({
+            onStdout: logger,
+            onStderr: logger,
+            rejectOnErrorCode: true,
+        })
+
+        return {
+            hostname: selection.instanceId,
+            envProvider,
+            sshPath: ssh,
+            vscPath: vsc,
+            SessionProcess,
+            selection,
+        }
+    }
+
+    private configureRemoteConnectionLogger(instanceId: string) {
+        const logPrefix = `ec2 (${instanceId})`
+        const logger = (data: string) => getLogger().verbose(`${logPrefix}: ${data}`)
+        return logger
+    }
+
+    public async configureSshKeys(selection: Ec2Selection, remoteUser: string): Promise<string> {
+        const keyPath = path.join(globals.context.globalStorageUri.fsPath, `aws-ec2-key`)
+        const keyPair = await SshKeyPair.getSshKeyPair(keyPath)
+        await this.sendSshKeyToInstance(selection, keyPair, remoteUser)
+        return keyPath
+    }
+
+    public async sendSshKeyToInstance(
+        selection: Ec2Selection,
+        sshKeyPair: SshKeyPair,
+        remoteUser: string
+    ): Promise<void> {
+        const sshKey = await sshKeyPair.getPublicKey()
+
+        const remoteAuthorizedKeysPaths = `/home/${remoteUser}/.ssh/authorized_keys`
+        const command = `echo "${sshKey}" > ${remoteAuthorizedKeysPaths}`
+        const documentName = 'AWS-RunShellScript'
+
+        await this.ssmClient.sendCommandAndWait(selection.instanceId, documentName, {
+            commands: [command],
+        })
+    }
+
+    protected async getRemoteUser(instanceId: string) {
+        const osName = await this.ssmClient.getTargetPlatformName(instanceId)
+        if (osName === 'Amazon Linux') {
+            return 'ec2-user'
```

Comment:
Does EC2/SSM have an API that could give this info?

Reply:
I don't believe so. There are a few ways of getting this information, but this method from SSM is what I found to be the most reliable. The defaults are listed in a table here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/TroubleshootingInstancesConnecting.html